Cancel package

zorkow opened the pull request:
Port of the MJ2 cancel package.

dpvc commented on mathjax3-ts/input/tex/ParseUtil.ts:
```diff
@@ -488,6 +488,30 @@ namespace ParseUtil {
+      const attr = attrib.replace(/ /g, '').split(/,/);
+      for (let i = 0, m = attr.length; i < m; i++) {
+        const keyvalue = attr[i].split(/[:=]/);
+        if (allowed[keyvalue[0]]) {
```
This should probably be `allowed.hasOwnProperty(keyvalue[0])`, since otherwise something like `[toString=1]` would be allowed.

zorkow replied:
Outdated.

dpvc commented on mathjax3-ts/input/tex/ParseUtil.ts:
```diff
@@ -488,6 +488,30 @@ namespace ParseUtil {
+  export function splitPackageOptions(attrib: string, allowed: {[key: string]: number} = {}): EnvList {
+    let def: EnvList = {};
+    if (attrib !== '') {
+      const attr = attrib.replace(/ /g, '').split(/,/);
```
I had occasion recently to need a function like this in a v2 extension, but the values that can be assigned could contain braced groups (so that you could include commas and equal signs, for example).  It might be good to handle that setup (which unfortunately means you can't just use splitting).  Alternatively, we could do another function when that is needed.

zorkow replied:
I've now implemented the full functionality of the LaTeX `keyval` function (https://www.ctan.org/pkg/keyval), with the exception that we remove arbitrary pairs of outer braces, while `keyval` only removes two pairs. But that should not make a difference.

dpvc commented on mathjax3-ts/input/tex/ParseUtil.ts:
```diff
@@ -488,6 +488,30 @@ namespace ParseUtil {
+    if (attrib !== '') {
+      const attr = attrib.replace(/ /g, '').split(/,/);
+      for (let i = 0, m = attr.length; i < m; i++) {
+        const keyvalue = attr[i].split(/[:=]/);
```
If you had `key=x:y`, then you will end up with `{key: 'x'}` in the end, and the `:y` will be lost.  I'm not sure what the best alternative is.

zorkow replied:
Outdated.

dpvc commented on mathjax3-ts/input/tex/cancel/CancelConfiguration.ts:
```diff
@@ -0,0 +1,98 @@
+  def ['notation'] = TexConstant.Notation.UPDIAGONALSTRIKE + ' ' +
+    TexConstant.Notation.UPDIAGONALARROW;
+  value = parser.create('node', 'mpadded', [value],
+                        {depth: '-.1em', height: ' + .1em', voffset: '.1em'});
```
I don't think you need the spaces in the `height` (and I think they might cause problems).

zorkow replied:
Space resulted from query - replace operation. Removed.

dpvc commented on mathjax3-ts/input/tex/cancel/CancelConfiguration.ts:
```diff
@@ -0,0 +1,98 @@
+
+
+/**
+ * The attributes allowed in \enclose{notation}[attributes]{math}
```
Is this supposed to refer to `\canel` rather than `\enclose`?

zorkow replied:
Actually no, as it accepts the same parameters as the `\enclose` macro. But the order is wrong.

dpvc commented on mathjax3-ts/input/tex/cancel/CancelConfiguration.ts:
```diff
@@ -0,0 +1,98 @@
+
+
+/**
+ * Parse function for cancel macros of the form \(b|x)cancel[attributes]{math}
```
Should `\(b|x)` be `\(b|x)?` or `\(b|x|)`?

dpvc commented on mathjax3-ts/input/tex/enclose/EncloseConfiguration.ts:
```diff
@@ -0,0 +1,77 @@
+  background: 1, mathbackground: 1,
+  padding: 1,
+  thickness: 1
+};
```
Can these be exported and used in the cancel file above, rather than having two copies?

zorkow replied:
Done.

dpvc commented on mathjax3-ts/input/tex/enclose/EncloseConfiguration.ts:
```diff
@@ -0,0 +1,77 @@
+};
+
+
+new CommandMap('enclose', {enclose:   'Enclose'}, EncloseMethods);
```
Remove extra spacing before `'Enclose'`?

zorkow commented:
* Actually the PR contains both the `cancel` and the `enclose` package. The former depends on the latter.
* I've now implemented the full functionality of the LaTeX `keyval` function (https://www.ctan.org/pkg/keyval), in `ParseUtil` to retrieve options. This should help with the `empheq` package as well.
* Added tests for `enclose` package.
* Added tests for attributes in `\cancel` and `\cancelto` macros.

dpvc commented:
Let me know when you want me to review this again.

zorkow commented:
Yes, please review again.

dpvc commented on mathjax3-ts/input/tex/ParseUtil.ts:
```diff
@@ -488,6 +488,110 @@ namespace ParseUtil {
+    let def: EnvList = readKeyval(attrib);
+    if (allowed) {
+      for (let key of Object.keys(def)) {
+        if (!allowed[key]) {
```
This still needs to be 

```suggestion
        if (!allowed.hasOwnProperty(key)) {
```

since otherwise it always allows `toString`, `toSource`, valueOf` or any of the other standard properties of an `Object` to be specified, which we don't want.

dpvc commented on mathjax3-ts/input/tex/ParseUtil.ts:
```diff
@@ -488,6 +488,110 @@ namespace ParseUtil {
+        if (!allowed[key]) {
+          if (error) {
+            throw new TexError('InvalidOption',
+                               'Invalid optional argument: ', key);
```
I think this wants to be

```suggestion
                               'Invalid optional argument: %1', key);
```

so that the key's name will be included in the message.

dpvc commented on mathjax3-ts/input/tex/ParseUtil.ts:
```diff
@@ -488,6 +488,110 @@ namespace ParseUtil {
+   * @return {string} The cleaned string.
+   */
+  function removeBraces(text: string): string {
+    while (text.match(/^{.*}$/)) {
```
I think there is a slight problem with this match.  For example, it will remove the braces from `a={x}and{y}` to give `a` the value `x}and{y`, which I don't think is what is wanted.  The braces should only be removed if they actually match.  (I haven't checked this with actual LaTeX, but I certainly *hop* it doesn't remove those braces.)

zorkow replied:
This is now handled correctly. I've added unit tests for the method.

dpvc commented on mathjax3-ts/input/tex/ParseUtil.ts:
```diff
@@ -488,6 +488,110 @@ namespace ParseUtil {
+    let value = '';
+    while (rest) {
+      let c = rest[0];
+      rest = rest.slice(1);
```
Making all these new copies off the string (one for each character) seems a bit inefficient.  Can you just use an index to move through the string rather than creating new strings each time you move to a new character?

zorkow replied:
Uses `index` counter now.

dpvc commented on mathjax3-ts/input/tex/cancel/CancelConfiguration.ts:
```diff
@@ -0,0 +1,87 @@
+  const math = parser.ParseArg(name);
+  const def = ParseUtil.keyvalOptions(attr, ENCLOSE_OPTIONS);
+  def ['notation'] = TexConstant.Notation.UPDIAGONALSTRIKE + ' ' +
+    TexConstant.Notation.UPDIAGONALARROW;
```
In the old days, MathML didn't have a notation for the arrow used by `\cancel`, and so MathJax used the (non-standard) `updiagonalarrow` here (and the output jax implemented it).  but since other renderers didn't know about it, we added `updiagonalstrike` as well so that if used in native MathML, there would be at least some indication of the arrow.

Since then, MathML has added a number of arrows, and I believe the correct one is `northeastarrow`, so we should change that here (it may need to be added to `TexConstant.Notation`).  It has been long enough that I would be happy to remove the `updiagonalstrike` and just use `northeastarrow`, but I'm OK if you want to leave it for backward compatibility.

zorkow replied:
For the record, official MathML notation is `longdiv, actuarial, phasorangle, radical, box, roundedbox, circle, left, right, top, bottom, updiagonalstrike, downdiagonalstrike, verticalstrike, horizontalstrike, northeastarrow, madruwb, text`.

zorkow replied:
OK, total confusion.... 
https://www.w3.org/TR/MathML2/chapter3.html#presm.menclose
gives has `northeastarrow` but no `updiagonalarrow`.

This site
https://www.tutorialspoint.com/mathml/mathml_enclosing.htm
(first hit on DuckDuckGo when including `northeastarrow` in search) has `northeastarrow` but no `updiagonalarrow` and a peculiar `text` element.

Finally the Moziall docs (first hit on DDG with only menclose:
https://developer.mozilla.org/en-US/docs/Web/MathML/Element/menclose
has `updiagonalarrow` but not `northeastarrow`.

Once again MathML is not a standard but a mess. 

We should retain `updiagonalarrow`.

dpvc commented on mathjax3-ts/input/tex/enclose/EncloseConfiguration.ts:
```diff
@@ -0,0 +1,79 @@
+  if (def.arrow) {
+    def.notation += ' updiagonalarrow';
+    delete def.arrow;
+  }
```
I think this is the same as my comments about `updiagonalarrow` as above.  It should be `northeastarrow`, and since `arrow` is not an allowed option (it's not in `ENCLOSE_OPTIONS`), I don't think this `if-then` will ever be triggered, in any case.  I think you can safely remove it.

We should probably keep a list of these changes from v2 so that they can be documented later.  I've started one, so if you have any others that you know of, let me know.

zorkow replied:
Removed.

zorkow commented:
PTAL.

Made the requested changes. In particular:
* Uses index for parsing in `keyval`.
* Correctly (but still greedily) removes outer braces and trims spaces.
* Changes to `notation` to `northeastarrow` and updates the corresponding enumerate in `TexConstants.ts`.

dpvc commented on mathjax3-ts/input/tex/ParseUtil.ts:
```diff
@@ -488,6 +488,123 @@ namespace ParseUtil {
+          start++;
+        } else if (start > braces) {
+          start = braces;
+        }
```
There is still a small problem with this function.  It will return `}` when the input is `{{}x}` because `start` will be 2 so two pairs of terminal characters are removed, leaving only the `}`.  That is because the `}` does end up setting `startCount` to `false`, but `start is never reduced to compensate for the `}` (since the `{` is never run again).

Instead, I think you want

```
      case ' ':
        break;
      case '{':
        braces++;
        if (startCount) {
          start = braces;
        }
        break;
      case '}':
        if (braces) {
          braces--;
        }
        if (startCount) {
            start = braces;
        }
      default:
        startCount = false;
      }
```

for the cases.  This make `start` be based on the brace count when `startCount` is turned off.  (I also like that it is a bit more symmetric.)

zorkow replied:
And I thought I had caught all corner cases... Thanks! Will correct and merge.

zorkow replied:
The problem I see with your solution is that you again only get the count for the balanced outer braces. Thus we get the same result as my initial, incorrect solutions.
But I feel we are nearly there.

dpvc replied:
Ah, right.  Sorry about that.  Perhaps something that include the `if-then` in the `}` case, but keeps the `start > braces` check in the `{` case?

zorkow replied:
I think I have a solution now. It's a bit ugly, though.

zorkow replied:
It uses two flags now: `startCount` to note that we are counting left outer brackets. `stopCount` to note that we are counting initial closing brackets. `startCount` flips on the first non-left-brace. `stopCount` is set after the first `}` and unset again after the first non-`}` after the initial `}`.
(If that makes sense.)

zorkow replied:
I'll intersperse comments into the code to explain my thinking.

Also, this is a typical automata theory problem. I'll run it by some colleagues one of these days. There must be a more elegant solution!

zorkow replied:
Still one case missing: `{}` is not removed.

dpvc replied:
I think if you just include

```
if (start > braces) {
  start = braces
}
```

in the `default` case of your original version, that would do it.  (I haven't looked at your new version yet).

dpvc replied:
Also, you would need to add

```
startCount = false;
break;
```
after the '}' case, since you don't want it to fall through to `default` after this change.

zorkow replied:
Thanks, but I think I got it now. Should at least work for now for all non-pathological cases. 
Unless you want to have another look, I'll merge.
